<a href="https://colab.research.google.com/github/alexseimakar/AI-systems-Homework/blob/main/%D0%9C%D0%B0%D0%BA%D0%B0%D1%80%D0%BE%D0%B2_%D0%90%D0%BB%D0%B5%D0%BA%D1%81%D0%B5%D0%B9_hw03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание №3

Задание выполнил(а): _(впишите свои фамилию и имя)_

### Общая информация

__Дата выдачи:__ 25.11.2024

__Дедлайн:__ 07.12.2024 23:59

### О задании

Часть 1 содержит теоретические задачи на линейные методы классификации и метрики качества классификации.

Часть 2 содержит практические задания на применение линейных методов классификации, подсчет метрик качества классификации и анализ результатов.

Обратите внимание, что вам не только нужно написать код, но и в некоторых местах ответить на вопросы.

### Оценивание и штрафы

Каждое задание оценивается одним баллом.

Оценка за ДЗ вычисляется по следующей формуле:

$$
\text{points} \times 10 / 16.8,
$$

где points — количество баллов, которое вы набрали. Всего в данном домашнем задании можно набрать 21 балл. Итого, максимальная оценка за данное домашнее задание - $21 \times 10 / 16.8 = 12.5$.

За сдачу задания позже срока на итоговую оценку за задание накладывается штраф в размере 0.25 балла в день, но получить отрицательную оценку нельзя.

__Внимание!__ Домашнее задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов.


# Часть 1 (10 баллов)

## 1

Какой из представленных графиков соответствует графику функции потерь в логистической регрессии? Тот же вопрос про метод опорных векторов.

**Ответ:**
- Логистическая регрессия: график № 1
- Метод опорных векторов: график № 3

## 2

При решении задачи бинарной классификации алгоритм на отложенной выборке верно предсказывает 100 из 400 объектов класса 0 и 75 из 100 объектов класса 1. Найдите долю правильных ответов данного классификатора.

**Ответ:** 0,35

## 3

Точность алгоритма классификации равна 0.6, а полнота равна 0.9. Чему в таком случае равно значение F-меры?

**Ответ:** 0,72

## 4

При решении задачи бинарной классификации использовались данные с двумя признаками. На данных была обучена логистическая регрессия. Известно, что после обучения вектор весов $(w_0, w_1, w_2) = (0.2, 0.7, 0.1)$ (вес $w_0$ соответствует единичному признаку). Для объекта $x$ с признаковым описанием $(3, 5)$ посчитайте модуль отступа.

_Не перепутайте с отступом классификатора - шириной разделяющей гиперплоскости. Это термин, более применимый к SVM. А отступ объекта (то, что требуется тут) - это немного другое, он показывает, ошибается ли классификатор на данном объекте и насколько._

**Ответ:** 2,8

## 5

При решении задачи бинарной классификации использовались данные с двумя признаками. На данных была обучена логистическая регрессия. Известно, что после обучения вектор весов $(w_0, w_1, w_2) = (0.2, 0.7, 0.1)$ (вес $w_0$ соответствует единичному признаку). Для объекта $x$ с признаковым описанием $(3, 5)$ вычислите предсказание вероятности принадлежности классу 1. Ответ округлите до двух знаков после запятой.

**Ответ:** 0,94

## 6

Какой из представленных графиков соответствует обученной логистической регрессии, для которой не настраивался свободный коэффициент - то есть $w_0 = 0$ (в `sklearn` это гиперпараметр `fit_intercept = False`)?

![logregs.png](attachment:logregs.png)

**Ответ:** график № 2

## 7

В задаче имеется 100 объектов. У каждого объекта 20 признаков. Сколько параметров (весов) будет настроено логистической регрессией в процессе обучения, учитывая свободный коэффициент?

**Ответ:** 21

## 8

Вы решаете задачу бинарной линейной классификации в трехмерном признаковом пространстве. Соответственно, решающее правило имеет вид:

$$
a(x) = \text{sign}(w_0 + w_1x_1 + w_2x_2 + w_3x_3)
$$

У вас есть следующая выборка:

| $x_1$ | $x_2$ | $x_3$ | $y$ |
|------|------|------|------|
| 0.2 | 0.4 | 0 | +1|
| 0.5 | 0.9 | 0 | +1|
| 0.3 | 0.3 | 0 | +1|
| 0.1 | 0.8 | 1 | -1|
| 0.5 | 0.7 | 1 | -1|
| 0.9 | 0.9 | 1 | -1|
| 0.1 | 0.3 | 1 | -1|

а) Сколько существует наборов коэффициентов $w = (w_0, w_1, w_2, w_3)$, при которых задача будет решена идеально (с accuracy = 1)?

б) Если ответ в п. а) - 0, пропустите это задание. Если ответ в п. а) - 1 или 2, запишите этот набор коэффициентов (или два набора соответственно). Иначе приведите хотя бы три таких набора коэффициентов.

**Ответ:**

а) Существует бесконечно много наборов коэффициентов, так как можно варьировать значения w0,w1,w2,w3​ в зависимости от условий

б) w0=1 w1=1 w2=1 w3=-3

## 9

Петя на работе построил алгоритм бинарной классификации $a(x)$, AUC-ROC которого равен 0.96. После этого Петя пошел обедать, а его коллега Вася, решив пошутить над Петей, сел за его компьютер и изменил код алгоритма так, чтобы вместо положительного класса алгоритм всегда предсказывал отрицательный, и наоборот – вместо отрицательного положительный. Какое значение AUC-ROC алгоритма обнаружил Вася, вернувшись с обеда?

**Ответ:** 0,04

## 10

Александр построил алгоритм бинарной классификации $a(x)$, AUC-ROC которого равен 0.66. Михаил построил алгоритм $b(x) = \frac{a(x)}{3}$. То есть для каждого объекта алгоритм $b(x)$ выдает вероятность, которая в три раза меньше, чем вероятность, выданная алгоритмом $a(x)$. Чему равно значение AUC-ROC алгоритма $b(x)$?

**Ответ:** 0,66

# Часть 2 (11 баллов)

![meme2.jpg](attachment:meme2.jpg)

In [1]:
import numpy as np
import pandas as pd

In [2]:
# устанавливаем актуальную версию sklearn
!pip install scikit-learn==0.23.2 --user

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.2/7.2 MB 42.3 MB/s eta 0:00:00
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Installing build dependencies ... error
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


В этом задании вы будете работать с [данными о пациентах, у части которых есть заболевание сердца](https://www.kaggle.com/ronitf/heart-disease-uci). Вам нужно построить модель, классифицирующую пациентов на больных этим заболеванием и тех, у кого его нет, а также проанализировать результаты.

In [3]:
data = pd.read_csv('heart.csv')
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Пациенты описываются следующими признаками:

- age
- sex
- chest pain type (4 values)
- resting blood pressure
- serum cholestoral in mg/dl
- fasting blood sugar > 120 mg/dl
- resting electrocardiographic results (values 0,1,2)
- maximum heart rate achieved
- exercise induced angina
- oldpeak = ST depression induced by exercise relative to rest
- the slope of the peak exercise ST segment
- number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

Целевая переменная:

- target (0 or 1)

## 1

Какой процент пациентов из представленных в данных имеет заболевание сердца (`'target' = 1`)?

In [4]:
len(data[data['target'] == 1]) / len(data) * 100

54.45544554455446

-----

Разделим данную выборку случайным образом на обучающую и тестовую части в отношении 3:1.

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data.drop('target', axis=1), data['target'], test_size=0.25, random_state=13)
X_train.shape, X_test.shape

((227, 13), (76, 13))

## 2

Обучите логистическую регрессию из `sklearn` (`sklearn.linear_model.LogisticRegression`) со следующими гиперпараметрами:

   - максимальное число итераций: `max_iter=1000`
   - без регуляризации: `penalty='none'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию

Подробнее о параметрах можно почитать в [документации](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

Какое значение свободного коэффициента у полученной логистической регрессии? А какое получилось значение остальных весов?

In [8]:
from sklearn.linear_model import LogisticRegression

model_logreg = LogisticRegression(max_iter=1000, penalty=None, random_state=13).fit(X_train, y_train)
model_logreg.intercept_,  model_logreg.coef_

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(array([4.02178791]),
 array([[-0.03278359, -1.87142394,  0.7468452 , -0.0083001 , -0.00338193,
          0.27996018,  0.50003437,  0.01749402, -1.08730588, -0.45883993,
          0.82636629, -0.75850861, -1.01954842]]))

## 3

Какое значение L2-нормы вектора весов (без учета свободного коэффициента) у полученного линейного классификатора?

_Напоминание. L2-норма вектора $v = (v_1, \ldots, v_n)$ - это корень из суммы квадратов его элементов:_

$$
\|v\|_2 = \sqrt{\sum\limits_{i=1}^nv_i^2}
$$

In [9]:
np.linalg.norm(model_logreg.coef_)

2.8425942737284804

## 4

Найдите долю правильных ответов (accuracy) классификатора на тестовой части выборки.

In [10]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test, model_logreg.predict(X_test))

0.8157894736842105

## 5

В задаче классификации, как и в задаче регрессии, для оптимизации линейных моделей можно применять регуляризацию. Этот метод реализован и в `sklearn.linear_model.LogisticRegression` - гиперпараметр, отвечающий за регуляризацию - это `C`, принимающее положительные значения. Класс реализован таким образом, что `C` является обратным значением к коэффициенту регуляризации - то есть чем меньше `C`, тем сильнее регуляризация, и наоборот. За тип регуляризации (нет регуляризации, L1, L2 или обе сразу) отвечает параметр `penalty`. Также для справки - есть гиперпараметр `solver`, задающий алгоритм, решающий оптимизационную проблему. Такие алгоритмы умеют работать не со всеми типами регуляризации, так что если, например, вы хотите обучить логистическую регрессию с L1-регуляризацией, но оставите гиперпараметр `solver` равным по умолчанию (`'lbfgs'`), то у вас возникнет ошибка (можете проверить).

Обучите логистическую регрессию заново со следующими гиперпараметрами:
   
   - максимальное число итераций: `max_iter=1000`
   - с L1-регуляризацией: `penalty='l1'`
   - с коэффициентом `C=0.1`
   - алгоритм оптимизации: `solver='liblinear'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию
   
Отличается ли качество полученного классификатора от качества первого? Какая доля правильных ответов получается теперь на тестовой выборке?

In [11]:
model_logreg_l1 = LogisticRegression(
    max_iter=1000,
    penalty='l1',
    random_state=13,
    C=0.1,
    solver='liblinear'
).fit(X_train, y_train)
accuracy_score(y_test, model_logreg_l1.predict(X_test))

0.8289473684210527

## 6

Выведите значения весов новой обученной логистической регрессии и сравните их с весами изначальной логистической регрессии. Что вы наблюдаете?

- Веса занулились

Найдите L2-норму вектора весов (без учета свободного коэффициента) для полученного классификатора. Как на нее повлияла регуляризация?

- Уменьшила её

In [12]:
np.linalg.norm(model_logreg_l1.coef_)

0.9158146328958144

## 7

Построим третью модель - метод опорных векторов (SVM). В `sklearn` этот метод реализован в классе [`sklearn.svm.SVC`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html). Мы проходили только линейную версию SVM (где мы рисуем гиперплоскость, разделяющую два класса) - ей соответствует гиперпараметр `kernel='linear'`.

Обучите SVM со следующими гиперпараметрами:

   - с коэффициентом `C=1`
   - с линейным ядром: `kernel='linear'`
   - сид: `random_state=13`
   - остальные гиперпараметры оставьте равными по умолчанию

Посчитайте долю правильных ответов построенного классификатора на тестовой выборке. Отличается ли оно от результатов двух построенных логистических регрессий?

- Да, их стало больше

In [13]:
from sklearn.svm import SVC

model_svc = SVC(C=1, kernel='linear', random_state=13).fit(X_train, y_train)
accuracy_score(y_test, model_svc.predict(X_test))

0.8421052631578947

## 8

Наконец, проверьте, как полученные классификаторы предсказывают не классы, а вероятности классов. В принципе, не все классификаторы позволяют предсказывать вероятности классов, но логистическая регрессия умеет это делать. Посмотрите на вероятности, которые выдает первая логистическая регрессия (обученный без регуляризации) на тестовой части выборки. В этом вам поможет метод [`predict_proba`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict_proba). Результатом его работы будет список размера $\ell\times 2$, где $\ell$ - это число объектов. В каждом столбце списка находятся вероятности соответствующего класса для объектов. Поэтому если вам нужен положительный класс, вас интересует последний столбец.

Какое получается значение AUC-ROC?

In [14]:
from sklearn.metrics import roc_auc_score

roc_auc_score(y_test, model_logreg.predict_proba(X_test)[:, 1])

0.8949964763918252

## 9

Посчитайте значение AUC-ROC для второй логистической регрессии (которая была обучена с L1-регуляризацией). Отличается ли оно от AUC-ROC первой логистической регрессии?

- Да, стало больше

In [15]:
roc_auc_score(y_test, model_logreg_l1.predict_proba(X_test)[:, 1])

0.9020436927413671

## 10

Наконец, посчитаем значение AUC-ROC для метода опорных векторов. По умолчанию SVM не умеет выдавать вероятности. Однако вспомните, что значение AUC-ROC зависит только от того, как отранжированы объекты - но не от самих значений, выдаваемых классификатором. Поэтому на самом деле AUC-ROC необязательно подсчитывать только для вероятностей - его можно посчитать, даже если классификатор выдает вещественные числа не из интервала $[0, 1]$.

В случае SVM мы можем посчитать расстояние от объектов до разделяющей гиперплоскости (со знаком минус, если объект относится к отрицательному классу, и плюс - если к положительному). В `sklearn` это реализовано с помощью метода [`decision_function`](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC.decision_function). В отличие от метода `predict_proba` в логистической регрессии, данный метод выдает не массив размера $\ell\times 2$, а массив из $\ell$ элементов.

Примените метод `decision_function` для уже обученного SVM на тестовой выборке. Посмотрите на то, что выдает этот метод.

Посчитайте значение AUC-ROC для полученного результата. Что в итоге в этой задаче сработало лучше - SVM или логистическая регрессия?

- SVM сработало чууууууууть-чуть лучше

In [16]:
roc_auc_score(y_test, model_svc.decision_function(X_test))

0.9034531360112755

## 11

Сопоставьте признаки в данных со значениями весов обученных моделей (метод `.coef_`). Для лучшей модели посчитайте, какой признак является самым важным для принятия решения (то есть имеет наибольший по модулю коэффициент). Выведите название соответствующего столбца.

In [17]:
print(f"Logreg: {data.columns[np.argmax(abs(model_logreg.coef_))]}")
print(f"Logreg: {data.columns[np.argmax(abs(model_logreg_l1.coef_))]}")
print(f"Logreg: {data.columns[np.argmax(abs(model_svc.coef_))]}")

Logreg: sex
Logreg: cp
Logreg: thal
